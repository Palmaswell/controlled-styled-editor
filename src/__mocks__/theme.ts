export const themeMock = {
  space: [0, 6, 12, 24, 48, 96, 192],
  fonts: {
    body: "'IBM Plex Serif', sans-serif",
    heading: "'IBM Plex Sans', 'Helvetica'",
  },
  fontSizes: [12, 18, 24, 30, 36, 48, 72],
  fontWeights: { body: 'normal', bold: 'bold', heading: 'bold' },
  lineHeights: { body: 1.5, heading: 1.1 },
  colors: {
    green: 'rgb(92, 185, 136)',
    yellows: ['rgb(255, 199, 84)', 'rgb(255, 202, 84)'],
    blues: ['rgb(99, 173, 242)', 'rgb(215, 236, 255)', 'rgb(77, 118, 164)'],
    reds: ['rgb(255, 92, 93)', 'rgb(243, 156, 165)'],
    blacks: ['rgb(0, 0, 0)', 'rgb(54, 54, 54)'],
    whites: ['rgb(255, 255, 255)', 'rgb(247, 247, 247)', 'rgb(244, 244, 244)'],
    grays: [
      'rgb(99, 99, 99)',
      'rgb(148, 148, 148)',
      'rgb(187, 187, 187)',
      'rgb(242, 242, 242)',
    ],
  },
  styles: {
    root: {
      fontFamily: 'body',
      fontSize: 2,
      fontWeight: 'body',
      lineHeight: 'body',
    },
    img: { padding: 0, margin: 0, marginBottom: 3, maxWidth: '100%' },
    h1: {
      padding: 0,
      margin: 0,
      marginBottom: 3,
      fontSize: 5,
      fontFamily: 'heading',
      lineHeight: 'heading',
      fontWeight: 'heading',
    },
    h2: {
      padding: 0,
      margin: 0,
      marginBottom: 3,
      fontSize: 4,
      fontFamily: 'heading',
      lineHeight: 'heading',
      fontWeight: 'heading',
    },
    h3: {
      padding: 0,
      margin: 0,
      marginBottom: 3,
      fontSize: 3,
      fontFamily: 'heading',
      lineHeight: 'heading',
      fontWeight: 'heading',
    },
    h4: {
      padding: 0,
      margin: 0,
      marginBottom: 3,
      fontSize: 2,
      fontFamily: 'heading',
      lineHeight: 'heading',
      fontWeight: 'heading',
    },
    h5: {
      padding: 0,
      margin: 0,
      marginBottom: 3,
      fontSize: 1,
      fontFamily: 'heading',
      lineHeight: 'heading',
      fontWeight: 'heading',
    },
    h6: {
      padding: 0,
      margin: 0,
      marginBottom: 3,
      fontSize: 0,
      fontFamily: 'heading',
      lineHeight: 'heading',
      fontWeight: 'heading',
    },
    ul: {
      padding: 0,
      margin: 0,
      marginBottom: 3,
      listStylePosition: 'outside',
      listStyleImage: 'none',
      ml: 3,
    },
    ol: {
      padding: 0,
      margin: 0,
      marginBottom: 3,
      listStylePosition: 'outside',
      listStyleImage: 'none',
      ml: 3,
    },
    li: {
      mb: 2,
      pl: 0,
      ol: { my: 2, ml: 3 },
      ul: { my: 2, ml: 3 },
      p: { mb: 2 },
    },
    p: { padding: 0, margin: 0, marginBottom: 3 },
    table: {
      padding: 0,
      margin: 0,
      marginBottom: 3,
      borderCollapse: 'collapse',
      width: '100%',
    },
    th: {
      textAlign: 'left',
      borderBottom: '1px solid',
      px: 2,
      py: 1,
      ':first-child': { pl: 0 },
      ':last-child': { pr: 0 },
    },
    td: {
      textAlign: 'left',
      borderBottom: '1px solid',
      px: 2,
      py: 1,
      mt: '-1px',
      ':first-child': { pl: 0 },
      ':last-child': { pr: 0 },
    },
    blockquote: { padding: 0, margin: 0, marginBottom: 3, mx: 3 },
    hr: {
      padding: 0,
      margin: 0,
      marginBottom: 3,
      border: 0,
      borderBottom: '1px solid',
      mt: '-1px',
      mb: 3,
    },
    b: { fontWeight: 'bold' },
    strong: { fontWeight: 'bold' },
    code: { fontSize: '85%' },
    pre: { padding: 3, margin: 0, marginBottom: 3, fontSize: '85%' },
  },
  borderWidths: [1, 2, 3, 4],
  letterSpacings: [0.5, 1, 1.5],
  spaces: [0, 3, 6, 12, 18, 24, 30, 36, 48, 72],
  transition: {
    duration: [0.33, 0.5, 0.66, 0.99],
    timing: [
      'cubic-bezier(0.455, 0.030, 0.515, 0.955)',
      'cubic-bezier(0.075, 0.820, 0.165, 1.000)',
      'cubic-bezier(1.000, 0.000, 0.000, 1.000)',
    ],
  },
  typography: {
    options: {
      baseFontSize: 16,
      baseLineHeight: 1.5,
      headerLineHeight: 1.1,
      scaleRatio: 2,
      googleFonts: [
        { name: 'IBM Plex Sans', styles: ['400', '600'] },
        { name: 'IBM Plex Serif', styles: ['400'] },
      ],
      headerFontFamily: ['IBM Plex Sans', 'Helvetica'],
      bodyFontFamily: ['IBM Plex Serif', 'sans-serif'],
      headerWeight: 'bold',
      bodyWeight: 'normal',
      boldWeight: 'bold',
      includeNormalize: true,
      blockMarginBottom: 1,
      rhythmUnit: 'px',
    },
  },
};
