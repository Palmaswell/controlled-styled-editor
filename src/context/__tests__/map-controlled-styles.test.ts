import * as ThemeQuery from 'theme-query';
import { SxStyleProp } from 'theme-ui';

import { mapControlledStyles, flatThemeTokens, getDefaultValue, mapDefaultValues } from '..';
import theme from '../../theme';

describe('Map Styles', () => {
  const qt = ThemeQuery.create({ theme, styles: 'object' });
  let colorsMock: string[] | undefined;
  let fontFamiliesMock: string[];
  let cssRuleMock: SxStyleProp;

  beforeAll(() => {
    colorsMock = [
      'rgb(92, 185, 136)',
      'rgb(255, 199, 84)',
      'rgb(255, 202, 84)',
      'rgb(99, 173, 242)',
      'rgb(215, 236, 255)',
      'rgb(77, 118, 164)',
      'rgb(255, 92, 93)',
      'rgb(243, 156, 165)',
      'rgb(0, 0, 0)',
      'rgb(54, 54, 54)',
      'rgb(255, 255, 255)',
      'rgb(247, 247, 247)',
      'rgb(244, 244, 244)',
      'rgb(99, 99, 99)',
      'rgb(148, 148, 148)',
      'rgb(187, 187, 187)',
      'rgb(242, 242, 242)',
      'currentColor',
    ];
    fontFamiliesMock = [
      "'IBM Plex Serif', sans-serif",
      "'IBM Plex Sans', 'Helvetica'",
    ];
    cssRuleMock = {
      padding: `${qt('spaces')(2)}px`,
      margin: `${qt('spaces')(5)}px`,
      color: qt('yellows')(1),
      backgroundColor: qt('yellows')(0),
      fontFamily: qt('heading'),
      fontSize: `${qt('fontSizes')(6)}px`,
    };
  });

  afterAll(() => {
    colorsMock = undefined;
    fontFamiliesMock = [''];
    cssRuleMock = {};
  });

  it('should return null if the sx argument is not an object', () => {
    expect(
      mapControlledStyles((undefined as unknown) as SxStyleProp, theme)
    ).toStrictEqual({});
    expect(
      mapControlledStyles((null as unknown) as SxStyleProp, theme)
    ).toStrictEqual({});
    expect(
      mapControlledStyles(([] as unknown) as SxStyleProp, theme)
    ).toStrictEqual({});
    expect(
      mapControlledStyles((1 as unknown) as SxStyleProp, theme)
    ).toStrictEqual({});
    expect(
      mapControlledStyles(('' as unknown) as SxStyleProp, theme)
    ).toStrictEqual({});
  });

  it('should return an empty object on an empty sx object', () => {
    expect(mapControlledStyles({}, theme)).toStrictEqual({});
  });

  it('should return an object with a property containing all the available font sizes', () => {
    expect(
      mapControlledStyles(
        {
          fontSize: `${qt('spaces')(2)}px`,
        },
        theme
      )
    ).toStrictEqual({ fontSize: theme.fontSizes });
  });

  it('should return an object with a property containing all the available font families', () => {
    expect(
      mapControlledStyles(
        {
          fontFamily: 'heading',
        },
        theme
      )
    ).toStrictEqual({ fontFamily: fontFamiliesMock });
  });

  it('should return an object with a property containing all available colors ', () => {
    expect(
      mapControlledStyles(
        {
          color: qt('yellows')(1),
        },
        theme
      )
    ).toStrictEqual({ color: colorsMock });
  });

  it('should return an object with a property containing all available colors ', () => {
    expect(
      mapControlledStyles(
        {
          backgroundColor: qt('yellows')(0),
        },
        theme
      )
    ).toStrictEqual({ backgroundColor: colorsMock });
  });

  it('should return an object with a property containing all spaces', () => {
    expect(
      mapControlledStyles(
        {
          padding: `${qt('spaces')(2)}px`,
        },
        theme
      )
    ).toStrictEqual({ padding: theme.spaces });
  });

  it('should return a filtered theme object mapped to CSS properties', () => {
    expect(mapControlledStyles(cssRuleMock, theme)).toStrictEqual({
      padding: theme.spaces,
      margin: theme.spaces,
      color: colorsMock,
      backgroundColor: colorsMock,
      fontFamily: fontFamiliesMock,
      fontSize: theme.fontSizes,
    });
  });

  describe('reduceThemeTokens()', () => {
    it('should return a flatten array with theme tokens', () => {
      expect(flatThemeTokens(qt('colors') as SxStyleProp)).toMatchSnapshot();
    });
  });

  describe('getDefaultValue()', () => {
    it('should return an empty string if no valid property is found', () => {
      expect(getDefaultValue('pointerEvents', cssRuleMock)).toBe('');
    });

    it('should return a rgb color value', () => {
      expect(getDefaultValue('color', cssRuleMock)).toBe(qt('yellows')(1));
      expect(getDefaultValue('backgroundColor', cssRuleMock)).toBe(
        qt('yellows')(0)
      );
    });

    it('should return a sanitized space scale value', () => {
      expect(getDefaultValue('padding', cssRuleMock)).toBe(
        qt('spaces')(2).toString()
      );
      expect(getDefaultValue('margin', cssRuleMock)).toBe(
        qt('spaces')(5).toString()
      );
      expect(
        getDefaultValue('padding', {
          padding: `${qt('spaces')(1)}px ${qt('spaces')(1)}px ${qt('spaces')(1)}px ${qt('spaces')(1)}px`,
        })
      ).toBe(
        `${qt('spaces')(1)} ${qt('spaces')(1)} ${qt('spaces')(1)} ${qt('spaces')(1)}`
      );
      expect(
        getDefaultValue('padding', {
          padding: `${qt('spaces')(6)}px ${qt('spaces')(0)}px ${qt('spaces')(2)}px`,
        })
      ).toBe(`${qt('spaces')(6)} ${qt('spaces')(0)} ${qt('spaces')(2)}`);
      expect(
        getDefaultValue('margin', {
          margin: `${qt('spaces')(3)}px ${qt('spaces')(5)}px`,
        })
      ).toBe(`${qt('spaces')(3)} ${qt('spaces')(5)}`);
    });

    it('should return a font family string', () => {
      expect(getDefaultValue('fontFamily', cssRuleMock)).toBe(fontFamiliesMock[1]);
    });

    it('should return a sanitized font size scale value', () => {
      expect(getDefaultValue('fontSize', cssRuleMock)).toBe(qt('fontSizes')(6).toString());
    });
  });

  describe('mapDefaultValues()', () => {
    it('should skip unknown properties', () => {
      expect(mapDefaultValues({ pointerEvents: 'all' })).toStrictEqual({});
      expect(mapDefaultValues({ backfaceVisibility: 'visible' })).toStrictEqual({});
    });

    it('should return a style object contining editor default values', () => {
      expect(mapDefaultValues(cssRuleMock)).toStrictEqual({
        padding: qt('spaces')(2).toString(),
        margin: qt('spaces')(5).toString(),
        color: qt('yellows')(1),
        backgroundColor: qt('yellows')(0),
        fontFamily: qt('heading'),
        fontSize: qt('fontSizes')(6).toString(),
      });
    });
  })

});
