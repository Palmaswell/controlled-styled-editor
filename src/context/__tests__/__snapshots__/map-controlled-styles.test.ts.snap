// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map Styles reduceThemeTokens() should return a flatten array with theme tokens 1`] = `
Array [
  "rgb(92, 185, 136)",
  "rgb(255, 199, 84)",
  "rgb(255, 202, 84)",
  "rgb(99, 173, 242)",
  "rgb(215, 236, 255)",
  "rgb(77, 118, 164)",
  "rgb(255, 92, 93)",
  "rgb(243, 156, 165)",
  "rgb(0, 0, 0)",
  "rgb(54, 54, 54)",
  "rgb(255, 255, 255)",
  "rgb(247, 247, 247)",
  "rgb(244, 244, 244)",
  "rgb(99, 99, 99)",
  "rgb(148, 148, 148)",
  "rgb(187, 187, 187)",
  "rgb(242, 242, 242)",
  "currentColor",
]
`;
